map_model/src/pathfind/v2.rs

```diff
@@ -0,0 +1,146 @@
+//! Structures related to the new road-based pathfinding
+//! (https://github.com/a-b-street/abstreet/issues/555) live here. When the transition is done,
+//! things here will probably move into pathfind/mod.rs.
+
+use anyhow::Result;
+
+use crate::{DirectedRoadID, Map, Path, PathRequest, PathStep, TurnID};
+
+/// Transform a sequence of roads representing a path into the current lane-based path, by picking
+/// particular lanes and turns to use.
+pub fn path_v2_to_v1(req: PathRequest, road_steps: Vec<DirectedRoadID>, map: &Map) -> Result<Path> {
+    // This is a somewhat brute-force method: run Dijkstra's algorithm on a graph of lanes and
+    // turns, but only build the graph along the path of roads we've already found. This handles
+    // arbitrary lookahead needed, and forces use of the original start/end lanes requested.
+    //
+    // Eventually we'll directly return road-based paths. Most callers will actually just use that
+    // directly, and mainly the simulation will need to expand to specific lanes, but it'll do so
+    // dynamically/lazily to account for current traffic conditions.
+    let mut graph = petgraph::graphmap::DiGraphMap::new();
+    for pair in road_steps.windows(2) {
+        for src in pair[0].lanes(req.constraints, map) {
+            for dst in pair[1].lanes(req.constraints, map) {
+                let turn = TurnID {
+                    parent: map.get_l(src).dst_i,
+                    src,
+                    dst,
+                };
+                if map.maybe_get_t(turn).is_some() {
+                    graph.add_edge(src, dst, turn);
+                }
+            }
+        }
+    }
+
+    match petgraph::algo::astar(
+        &graph,
+        req.start.lane(),
+        |l| l == req.end.lane(),
+        // TODO We could include the old lane-changing penalties here, but I'm not sure it's worth
+        // the complication. The simulation layer will end up tuning those anyway.
```

Comment:
Does this mean a regression in lane changing behavior once this is merged?

Reply:
Not yet, because the Dijkstra's pathfinding isn't ever used. You have to import with `--skip_ch` to switch to it.

But in the next step, which will change the CH implementation too, then this will introduce a regression. I believe it should only affect LCing near uber-turns, since the simulation layer doesn't `opportunistically_lanechange` near uber-turns. I'll add the costs back in during that PR.